File: GaitScheduler/lcm_data_exchange_gs.py
import lcm
import numpy as np
from transforms3d.euler import euler2mat

from mors_msgs.phase_signal_msg import phase_signal_msg
from mors_msgs.robot_state_msg import robot_state_msg
from mors_msgs.gait_params_msg import gait_params_msg
from mors_msgs.robot_cmd_msg import robot_cmd_msg

class LCMDataExchange():
    def __init__(self,
                 gait_params_cnannel = "GAIT_PARAMS",
                 gait_phase_channel = "GAIT_PHASE",
                 robot_state_channel = "ROBOT_STATE",
                 robot_cmd_channel = "ROBOT_CMD",
                 robot_ref_channel = "ROBOT_REF",
                 ) -> None:
        self.gait_params_cnannel = gait_params_cnannel
        self.gait_phase_channel = gait_phase_channel
        self.robot_state_channel = robot_state_channel
        self.robot_cmd_channel = robot_cmd_channel
        self.robot_ref_channel = robot_ref_channel

        self.gait_type = [np.pi, 0, 0, np.pi]
        self.t_st = 0.4
        self.t_sw = 0.25
        self.standing = True
        self.stride_height = 0.0

        self.gait_phase = [0]*4
        self.gait_phi = [0]*4

        self.ref_pos = [0.0]*3
        self.ref_orientation = [0.0]*3
        self.ref_lin_vel = [0.0]*3
        self.ref_ang_vel = [0.0]*3

        self.leg_contact_state = [0]*4
        self.body_pos = [0,0,0]
        self.body_rpy = [0.0]*3
        self.foot_pos_local = np.array([0.0]*4, dtype=np.float64)


        self.phase_sig_msg = phase_signal_msg()
        self.rbt_state_msg = robot_state_msg()
        self.gait_prms_msg = gait_params_msg()
        self.robot_ref_msg = robot_cmd_msg()
        self.robot_cmd_msg = robot_cmd_msg()
        self.lc_cmd = lcm.LCM()
        

    def robot_state_callback(self, channel, data : robot_state_msg):
        state_msg = robot_state_msg.decode(data)
        self.leg_contact_state = state_msg.legs.contact_states[:]
        self.body_rpy = state_msg.body.orientation[:]
        self.body_pos = state_msg.body.position[:]

        self.R_body = euler2mat(self.body_rpy[0], self.body_rpy[1], self.body_rpy[2])

        self.foot_pos_local = np.array([self.R_body @ np.array(state_msg.legs.r1_pos, dtype=np.float64).reshape(3,),
                                        self.R_body @ np.array(state_msg.legs.l1_pos, dtype=np.float64).reshape(3,),
                                        self.R_body @ np.array(state_msg.legs.r2_pos, dtype=np.float64).reshape(3,),
                                        self.R_body @ np.array(state_msg.legs.l2_pos, dtype=np.float64).reshape(3,)])


    def robot_state_thread(self):
        print("RobotState thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.robot_state_channel, self.robot_state_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass

    def gait_params_callback(self, channel, data : gait_params_msg):
        msg = gait_params_msg.decode(data)
        self.gait_type = msg.gait_type[:]
        self.t_st = msg.t_st
        self.t_sw = msg.t_sw
        self.standing = msg.standing
        self.stride_height = msg.stride_height


    def gait_params_thread(self):
        print("GaitParams thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.gait_params_cnannel, self.gait_params_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass

    def robot_ref_callback(self, channel, data : robot_cmd_msg):
        msg = robot_cmd_msg.decode(data)
        self.ref_pos = msg.cmd_pose[:3]
        self.ref_orientation = msg.cmd_pose[3:]
        self.ref_lin_vel = msg.cmd_vel[:3]
        self.ref_ang_vel = msg.cmd_vel[3:]


    def robot_ref_thread(self):
        print("GaitParams thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.robot_ref_channel, self.robot_ref_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass


    def send_gait_phase(self, phase, phi):
        self.phase_sig_msg.phase = phase[:]
        self.phase_sig_msg.phi = phi[:]

        self.lc_cmd.publish(self.gait_phase_channel, self.phase_sig_msg.encode())

    def send_robot_cmd(self, pos, orientation, lin_vel, ang_vel):
        self.robot_cmd_msg.cmd_pose[:3] = pos[:]
        self.robot_cmd_msg.cmd_pose[3:] = orientation[:]
        self.robot_cmd_msg.cmd_vel[:3] = lin_vel[:]
        self.robot_cmd_msg.cmd_vel[3:] = ang_vel[:]

        self.lc_cmd.publish(self.robot_cmd_channel, self.robot_cmd_msg.encode())


    def get_leg_contacts(self):
        return self.leg_contact_state
    
    def get_leg_pos_local(self):
        return self.foot_pos_local
    
    def get_body_state(self):
        return self.body_pos, self.body_rpy
    
    def get_gait_params(self):
        return self.gait_type, self.t_st, self.t_sw, self.standing
    
    def get_robot_ref(self):
        return self.ref_pos, self.ref_orientation, self.ref_lin_vel, self.ref_ang_vel

File: GaitScheduler/src/lcm_data_exchange_gs.cpp
#include "lcm_data_exchange_gs.hpp"
#include <unistd.h>

// #define KT 0.74
// #define GEAR_RATIO 10.0

LCMExchanger::LCMExchanger()
{
    
    if(!robot_ref_subscriber.good())
        return;
    if(!robot_state_subscriber.good())
        return;
    if(!gait_params_subscriber.good())
        return;

    string config_address = mors_sys::GetEnv("CONFIGPATH");//cwd;
    config_address += "/channels.yaml";

    YAML::Node channel_config = YAML::LoadFile(config_address);//"/home/user/mors_mpc_control/config/channels.yaml");
    robot_cmd_channel = channel_config["robot_cmd"].as<string>();
    robot_ref_channel = channel_config["robot_ref"].as<string>();
    robot_state_channel = channel_config["robot_state"].as<string>();
    phase_signal_channel = channel_config["gait_phase"].as<string>();
    gait_params_channel = channel_config["gait_params"].as<string>();
    
    // cmd_vel.resize(6);
    // cmd_pose.resize(6);
    // phase.resize(4);
    robot_state.ang_vel.resize(3);
    robot_state.lin_vel.resize(3);
    robot_state.orientation.resize(3);
    robot_state.pos.resize(3);
    robot_ref.ang_vel.resize(3);
    robot_ref.lin_vel.resize(3);
    robot_ref.orientation.resize(3);
    robot_ref.pos.resize(3);
    leg_state.contacts.resize(4);
    leg_state.r1_pos.resize(3);
    leg_state.l1_pos.resize(3);
    leg_state.r2_pos.resize(3);
    leg_state.l2_pos.resize(3);

    // cmd_vel.setZero();
    // cmd_pose.setZero();
    // phase << 1,1,1,1;
    robot_state.ang_vel.setZero();
    robot_state.lin_vel.setZero();
    robot_state.orientation.setZero();
    robot_state.pos.setZero();
    robot_ref.ang_vel.setZero();
    robot_ref.lin_vel.setZero();
    robot_ref.orientation.setZero();
    robot_ref.pos.setZero();
    leg_state.contacts = {true, true, true, true};
    leg_state.r1_pos.setZero();
    leg_state.l1_pos.setZero();
    leg_state.r2_pos.setZero();
    leg_state.l2_pos.setZero();
    t_sw = 0.25;
    t_st = 0.4;
    stride_height = 0.05;
    gait_type = {M_PI, 0, 0, M_PI};
    standing = false;
}

void LCMExchanger::start_exchanger()
{
    thRobotRef = make_unique<thread> (&LCMExchanger::robotRefThread, this);
    thRobotState = make_unique<thread> (&LCMExchanger::robotStateThread, this);
    thGaitParams = make_unique<thread> (&LCMExchanger::gaitParamsThread, this);
}

void LCMExchanger::robotRefHandler(const lcm::ReceiveBuffer* rbuf,
                                    const std::string& chan,
                                    const mors_msgs::robot_cmd_msg* msg)
{
    for (int i=0; i<3; i++)
    {
        // cmd_pose(i) = msg->cmd_pose[i];
        // cmd_vel(i) = msg->cmd_vel[i];
        robot_ref.pos(i) = msg->cmd_pose[i];
        robot_ref.orientation(i) = msg->cmd_pose[i+3];
        robot_ref.lin_vel(i) = msg->cmd_vel[i];
        robot_ref.ang_vel(i) = msg->cmd_vel[i+3];
    }
}

void LCMExchanger::robotStateHandler(const lcm::ReceiveBuffer* rbuf,
                                    const std::string& chan,
                                    const mors_msgs::robot_state_msg* msg)
{
    for (int i=0; i<3; i++)
    {
        leg_state.r1_pos(i) = msg->legs.r1_pos[i];
        leg_state.l1_pos(i) = msg->legs.l1_pos[i];
        leg_state.r2_pos(i) = msg->legs.r2_pos[i];
        leg_state.l2_pos(i) = msg->legs.l2_pos[i];

        // leg_state.r1_vel(i) = msg->legs.r1_vel[i];
        // leg_state.l1_vel(i) = msg->legs.l1_vel[i];
        // leg_state.r2_vel(i) = msg->legs.r2_vel[i];
        // leg_state.l2_vel(i) = msg->legs.l2_vel[i];

        // leg_state.r1_grf(i) = msg->legs.r1_grf[i];
        // leg_state.l1_grf(i) = msg->legs.l1_grf[i];
        // leg_state.r2_grf(i) = msg->legs.r2_grf[i];
        // leg_state.l2_grf(i) = msg->legs.l2_grf[i];

        leg_state.contacts[i] = msg->legs.contact_states[i];

        robot_state.pos(i) = msg->body.position[i];
        robot_state.orientation(i) = msg->body.orientation[i];
        robot_state.lin_vel(i) = msg->body.lin_vel[i];
        robot_state.ang_vel(i) = msg->body.ang_vel[i];
    }
    leg_state.contacts[3] = msg->legs.contact_states[3];
}

void LCMExchanger::gaitParamsHandler(const lcm::ReceiveBuffer* rbuf,
                                    const std::string& chan,
                                    const mors_msgs::gait_params_msg* msg)
{
    // for (int i=0; i<4; i++)
    //     phase(i) = msg->phase[i];
    t_sw = msg->t_sw;
    t_st = msg->t_st;
    stride_height = msg->stride_height;
    standing = msg->standing;
    for (int i = 0; i < 4; i++)
        gait_type[i] = msg->gait_type[i];
    // cout << phase.transpose() << endl;
}

void LCMExchanger::robotRefThread()
{   
    robot_ref_subscriber.subscribe(robot_ref_channel, &LCMExchanger::robotRefHandler, this);
    while(true)
    {
        robot_ref_subscriber.handle();
        // cout << "grf thread" << endl;
        // sleep(0.001);
    }
}

void LCMExchanger::robotStateThread()
{   
    robot_state_subscriber.subscribe(robot_state_channel, &LCMExchanger::robotStateHandler, this);
    while(true)
    {
        robot_state_subscriber.handle();
        // cout << "grf thread" << endl;
        // sleep(0.001);
    }
}

void LCMExchanger::gaitParamsThread()
{
    gait_params_subscriber.subscribe(gait_params_channel, &LCMExchanger::gaitParamsHandler, this);
    while(true)
    {
        gait_params_subscriber.handle();
        // cout << "imu thread" << endl;
        // sleep(0.001);
    }
}


void LCMExchanger::sendRobotCmd(RobotData& robot_cmd)
{
    for (int i=0; i<3; i++)
    {
        robotCmdMsg.cmd_pose[i] = robot_cmd.pos(i);
        robotCmdMsg.cmd_pose[i+3] = robot_cmd.orientation(i);
        robotCmdMsg.cmd_vel[i] = robot_cmd.lin_vel(i);
        robotCmdMsg.cmd_vel[i+3] = robot_cmd.ang_vel(i);
    }
    robot_cmd_publisher.publish(robot_cmd_channel, &robotCmdMsg);
}

void LCMExchanger::sendPhaseSig(vector<int>& phase, vector<double>& phi)
{
    for (int i = 0; i < 4; i++)
    {
        phaseSigMsg.phase[i] = phase[i];
        phaseSigMsg.phi[i] = phi[i];
    }
    phase_sig_publisher.publish(phase_signal_channel, &phaseSigMsg);
}

RobotData LCMExchanger::getRobotRef()
{
    return robot_ref;
}

RobotData LCMExchanger::getBodyState()
{
    return robot_state;
}

LegData LCMExchanger::getLegState()
{
    return leg_state;
}

void LCMExchanger::get_gait_params(double& t_st, double& t_sw, vector<double>& gait_type, bool& standing, double& stride_height)
{
    t_st = this->t_st;
    t_sw = this->t_sw;
    gait_type = this->gait_type;
    standing = this->standing;
    stride_height = this->stride_height;
}



LCMExchanger::~LCMExchanger()
{

}
